src/report/chunked.rs

```diff
@@ -0,0 +1,992 @@
+                    unsafe { MaybeUninit::uninit().assume_init() };
+                for (source, sink) in payload_bytes.iter().zip(payload.iter_mut()) {
+                    *sink = MaybeUninit::new(*source);
+                }
+                NativeChunk::Extension {
+                    header,
+                    contents: NativeChunkExtensionContents {
+                        n_chunk_payload_bytes: n_payload_bytes,
+                        payload,
+                    },
+                }
+            }
+            data_type_byte => {
+                return Err(ParseChunkFromWireError::UnsupportedDataType(data_type_byte))
+            }
+        })
+    }
+}
+
+/// Everything that can go wrong converting bytes off the wire to a NativeChunk
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParseChunkFromWireError {
+    /// The fingerprint didn't match expectations
+    InvalidFingerprint,
+    /// There weren't enough bytes for a full header
+    MissingHeader,
+    /// There weren't enough payload bytes (based on
+    /// expectations from inspecting the header).
+    IncompletePayload,
+    /// The data type was not one of the supported varieties
+    UnsupportedDataType(u8),
+    /// The tracer id didn't follow the rules for being
+    /// a valid ekotrace-location-specifying TracerId
+    InvalidTracerId(u32),
+}
+
+/// Chunk framing information, owned and accessible in the native
+/// endianness.
+#[derive(Debug, PartialEq)]
+pub struct NativeChunkHeader {
+    /// The tracer_id (a.k.a. location id) of the
+    /// ekotrace agent instance producing this report.
+    pub location_id: TracerId,
+    /// A u16 representing the group of report chunks to which
+    /// this chunk belongs. Determined by the ekotrace instance.
+    /// Expected, but not guaranteed to be increasing
+    /// and wrapping-overflowing during the lifetime of an ekotrace
+    /// instance.
+    pub chunk_group_id: u16,
+    /// In what ordinal position does this chunk's payload live relative to the other chunks.
+    pub chunk_index: u16,
+    /// Is this chunk the last chunk in the report (0001) or not (0000)?
+    pub is_last_chunk: bool,
+    /// Reserved for future enhancements and to make the payload 4-byte aligned
+    pub reserved: u8,
+}
+
+/// The contents of the chunk, interpreted as the compact log format
+///
+/// This struct maintains the invariant that the members of the `payload`
+/// field are initialized up to `n_chunk_payload_items` value
+pub struct NativeChunkLogContents {
+    /// How many of the payload items are populated?
+    n_chunk_payload_items: u8,
+    /// The content of the report chunk
+    payload: [MaybeUninit<CompactLogItem>; MAX_PAYLOAD_COMPACT_LOG_ITEMS_PER_CHUNK],
+}
+
+impl PartialEq for NativeChunkLogContents {
+    fn eq(&self, other: &Self) -> bool {
+        self.payload
+            .iter()
+            .take(usize::from(self.n_chunk_payload_items))
+            .zip(
```

Comment:
This will return a false positive if the two have different sizes, but one is a prefix of the other.

Reply:
Good catch, thank you!